### PR TITLE
transpose: add num_batches to batch independent transposes into one dispatch

### PR DESCRIPTION
**Pull request by atassis**
GEMV and StridedCopy already take `num_batches` to batch B independent same-shape operations into a single dispatch; Transpose did not, forcing callers to unroll B per-head/per-batch transposes into B separate dispatches for identical kernel work (a common multi-head-attention pattern).

**Added**
- `num_batches` on `Transpose` (default 1). `num_batches>1` lays B contiguous (M,N) matrices back-to-back and streams them through the same ObjectFifos (one task group per batch); the core still only sees s×s sub-tiles, so the kernel is unchanged.
- `num_batches>1` test coverage (the batched path was previously untested), with a batched golden reference.

**Changed**
- `get_arg_spec` prepends a batch dim only when `num_batches>1`; `num_batches=1` is byte-identical to the previous single-transpose schedule.

**Removed**
- None.

Verified on device (NPU2): `num_batches` in {1, 2, 4} pass. Mirrors GEMV's existing `num_batches`.

### CODE REVIEW

File: iron/operators/transpose/design.py [hunk 1]
```diff
@@ -47,13 +50,25 @@ def shuffle_transpose(dev, M, N, num_columns, num_channels, m, n, s, func_prefix
     # and channels. Partially transposes the input
     # data so that the kernel only needs to
     # transpose s*s-sized sub-tiles.
+    # For num_batches>1 the L3 tensors hold that many contiguous (M,N) matrices, stacked along
+    # the row dimension: in-dims (num_batches*M, N), out-dims (num_batches*N, M). At num_batches==1
+    # these reduce to (M,N)/(N,M) — identical to the original single-transpose patterns. Each (i,j)
+    # column/channel gets one TAP per batch (offset += batch*num_elements); the per-batch internal
+    # sizes/strides are unchanged because each matrix is contiguous and row-major.
```
Comment (andrej) [inline iron/operators/transpose/design.py]: Some of this comment reads a bit like a commit message / PR description, since it refers to the "original" implementation and "unchanged" things. This will be confusing when reading the code as-is once merged. Can you please reword this?

File: iron/operators/transpose/design.py [hunk 2]
```diff
@@ -134,29 +157,33 @@ def core_body(of_in1, of_out, transpose_kernel):
...
-                    of_outs[i * num_channels + j].cons(),
-                    C,
-                    taps_out_L1L3[i * num_channels + j],
-                    wait=True,  # wait for the transfer to complete and data to be available
-                    task_group=tg,
-                )
-        rt.finish_task_group(tg)
+        # One task group per batch (each a parallel fill+drain over all columns/channels), so the
+        # num_batches contiguous matrices stream through the same FIFOs in sequence. At num_batches==1
+        # this is a single pass — identical to the original single-transpose schedule.
```
Comment (andrej) [inline iron/operators/transpose/design.py]:
```suggestion
        # num_batches contiguous matrices stream through the same FIFOs in sequence.
```

File: iron/operators/transpose/design.py [hunk 3]
```diff
@@ -134,29 +157,33 @@ def core_body(of_in1, of_out, transpose_kernel):
...
+            # Fill the input objectFIFOs with data
+            for i in range(num_columns):
+                for j in range(num_channels):
+                    rt.fill(
+                        of_in1s_L3L2[i * num_channels + j].prod(),
+                        A,
+                        taps_in_L3L2[i * num_channels + j][batch],
+                        task_group=tg,
+                    )
+            # Drain the output objectFIFOs with data
```
Comment (andrej) [inline iron/operators/transpose/design.py]:
```suggestion
            # Drain the output objectFIFOs of data
```

**Review by jgmelber — APPROVED** [review]

File: iron/operators/transpose/op.py [hunk 4]
```diff
@@ -90,7 +104,10 @@ def get_kernel_artifacts(self):
         ]
 
     def get_arg_spec(self):
+        # num_batches==1 keeps the exact flat (M*N,) spec (unchanged behavior for existing callers);
+        # >1 prepends a batch dim over contiguous matrices.
```
Comment (andrej) [inline iron/operators/transpose/op.py]:
```suggestion
```

File: iron/operators/transpose/reference.py [hunk 5]
```diff
@@ -5,9 +5,21 @@
...
+        output_tensor = torch.transpose(input_tensor, 0, 1)
+        return {"input": input_tensor, "output": output_tensor}
+    # num_batches>1: B independent (rows,cols) matrices laid back-to-back; each is
+    # transposed independently and the results concatenated in the same order.
+    input_tensor = (
+        torch.rand(num_batches, rows, cols, dtype=torch_dtype_map[dtype]) * val_range
+    )
+    output_tensor = torch.stack(
+        [torch.transpose(input_tensor[b], 0, 1) for b in range(num_batches)]
+    )
```
Comment (andrej) [inline iron/operators/transpose/reference.py]:
```suggestion
    # num_batches>1: B independent (rows,cols) matrices laid back-to-back; each is
    # transposed independently and the results concatenated in the same order.
    input_tensor = (
        torch.rand(num_batches, rows, cols, dtype=torch_dtype_map[dtype]) * val_range
    )
    output_tensor = torch.stack(
        [torch.transpose(input_tensor[b], 0, 1) for b in range(num_batches)]
    )
    # drop batch dimension if num_batches == 1
    input_tensor = torch.squeeze(input_tensor, 0)
    output_tensor = torch.squeeze(output_tensor, 0)
```

reason for suggestion: deduplicate random tensor generation logic so we generate them in the same way for each test case

**Review by andrej — CHANGES_REQUESTED** [review]
Thank you for the contribution! This will be a useful addition. Just a couple nitpicks, then please rebase on `devel` and we can go ahead and merge this.

**Comment by andrej** [comment]
Note the CI failures we're seeing should disappear after a rebase. :)

**Comment by atassis** [comment]
Hi there! Thanks for the feedback, have applied it in both PRs.
I have also made a https://github.com/Xilinx/mlir-aie/pull/3178 PR with compile-time fix O(n^2) to O(1) in several places which haven't got a human attention, it seems. I'll be grateful for any help to move that forward. I have a few more generic fixes queued after that, if one proves to be useful =)
BTW, I am trying to build a general purpose multi precision NPU inference engine on IRON for consumer Ryzen AI / XDNA2 laptops (might later make it work with other hardware, too, though, I have none to test against), and this work helped me to find out these fixes.
So, not promising, but might make some more PR's in close future!

**Comment by thomthehound** [comment]
> Hi there! Thanks for the feedback, have applied it in both PRs. I have also made a [Xilinx/mlir-aie#3178](https://github.com/Xilinx/mlir-aie/pull/3178) PR with compile-time fix O(n^2) to O(1) in several places which haven't got a human attention, it seems. I'll be grateful for any help to move that forward. I have a few more generic fixes queued after that, if one proves to be useful =) BTW, I am trying to build a general purpose multi precision NPU inference engine on IRON for consumer Ryzen AI / XDNA2 laptops (might later make it work with other hardware, too, though, I have none to test against), and this work helped me to find out these fixes. So, not promising, but might make some more PR's in close future!

I think both of these contributions are valuable, so thank you for them! 

Your [Xilinx/mlir-aie#3178] PR has had human eyes on it, but, speaking for myself, I would prefer to see the CoPilot review issues there explained or resolved before commenting further.

**Comment by atassis** [comment]
@thomthehound good point, resolved them

**Review by andrej — APPROVED** [review]

**Comment by atassis** [comment]
@andrej hi there. A friendly reminder if we can merge this, so I can reduce the PR surface area. I have some more things in draft, still, to suggest, but don't want to spam with them and keep lots of open PRs at once

**Comment by andrej** [comment]
@atassis Sorry for the delay on this!

**Comment by atassis** [comment]
@andrej No trouble =)
More PR DDOS coming, I think
Glad to be useful, if I am
